### PR TITLE
fix(topic): performance for read and creation

### PR DESCRIPTION
**Pull request by ivan-savciuc**
## About this change—what it does

Fix Kafka Topic create and read performance

### CODE REVIEW

**Comment by byashimov** [comment]
@ivan-savciuc hey is this for v4 or v3, or both?

**Review by byashimov — DISMISSED** [review]
lgtm

File: internal/sdkprovider/service/kafkatopic/kafka_topic.go [hunk 1]
```diff
@@ -269,18 +269,11 @@ func resourceKafkaTopicCreate(ctx context.Context, d *schema.ResourceData, m int
 	// aiven.KafkaTopics.Create() function may return 501 on create
 	// Second call might say that topic already exists
 	// So to be sure, better check it before create
-	_, err := getTopic(ctx, m, d.Timeout(schema.TimeoutRead), project, serviceName, topicName)
-
-	// No error means topic exists
-	if err == nil {
+	client := m.(*aiven.Client)
+	if isTopicExists(ctx, client, project, serviceName, topicName) {
```
Comment (Serpentiel) [inline internal/sdkprovider/service/kafkatopic/kafka_topic.go:273]: consider adding a comment above this block to briefly explain its purpose for clarity

File: internal/sdkprovider/service/kafkatopic/kafka_topic_cache.go [hunk 2]
```diff
@@ -201,3 +201,11 @@ func DeleteTopicFromCache(projectName, serviceName, topicName string) {
 	}
 	t.Unlock()
 }
+
+// GetFullQueue retrieves a topics queue
+func (t *kafkaTopicCache) GetFullQueue(projectName, serviceName string) []string {
```
Comment (Serpentiel) [inline internal/sdkprovider/service/kafkatopic/kafka_topic_cache.go:206]: ensure the method name accurately represents its purpose; If the function retrieves a specific type of queue or has a unique behavior, be more explicit in the name

File: internal/sdkprovider/service/kafkatopic/kafka_topic_cache.go [hunk 3]
```diff
@@ -201,3 +201,11 @@ func DeleteTopicFromCache(projectName, serviceName, topicName string) {
...
+	t.RLock()
+	defer t.RUnlock()
+
+	return t.inQueue[projectName+serviceName]
```
Comment (Serpentiel) [inline internal/sdkprovider/service/kafkatopic/kafka_topic_cache.go:210]: consider handling a scenario where the key might not exist in the `inQueue` map; directly accessing a non-existent key will return a nil slice, which might be unexpected for the function caller

File: internal/sdkprovider/service/kafkatopic/kafka_topic_exists.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import (
+	"context"
+	"log"
+	"sync"
+
+	"github.com/aiven/aiven-go-client/v2"
+	"golang.org/x/exp/slices"
+)
+
+var onceForService sync.Map
```
Comment (Serpentiel) [inline internal/sdkprovider/service/kafkatopic/kafka_topic_exists.go:12]: consider renaming the variable to be more descriptive, like `onceCheckTopicForService` or something similar

File: internal/sdkprovider/service/kafkatopic/kafka_topic_exists.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+// isTopicExists checks if topic exists
+func isTopicExists(ctx context.Context, client *aiven.Client, project, serviceName, topic string) bool {
+	c := getTopicCache()
+
+	var err error
+	once, _ := onceForService.LoadOrStore(project+"/"+serviceName, new(sync.Once))
+	once.(*sync.Once).Do(func() {
```
Comment (Serpentiel) [inline internal/sdkprovider/service/kafkatopic/kafka_topic_exists.go:20]: add comments for better clarity, especially on the significance of using `sync.Once`

File: internal/sdkprovider/service/kafkatopic/kafka_topic_exists.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+		var list []*aiven.KafkaListTopic
+		list, err = client.KafkaTopics.List(ctx, project, serviceName)
+		if err != nil {
+			return
+		}
+
+		c.SetV1List(project, serviceName, list)
+	})
+
+	if err != nil {
```
Comment (Serpentiel) [inline internal/sdkprovider/service/kafkatopic/kafka_topic_exists.go:30]: this approach can be confusing: if possible, consider handling the error within the `once.Do` block, or refactoring the approach to make the flow clearer

File: internal/sdkprovider/service/kafkatopic/kafka_topic_exists.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+		log.Printf("[ERROR] cannot check kafka topic existence: %s", err)
+		return false
+	}
+
+	if slices.Contains(c.GetV1List(project, serviceName), topic) || slices.Contains(c.GetFullQueue(project, serviceName), topic) {
```
Comment (Serpentiel) [inline internal/sdkprovider/service/kafkatopic/kafka_topic_exists.go:35]: for better performance, if the topic is found in the first `c.GetV1List` check, there's no need to check in the second `c.GetFullQueue` call: you can return `true` immediately

**Review by Serpentiel — CHANGES_REQUESTED** [review]

**Comment by ivan-savciuc** [comment]
Converting this patch for v4 instead of v5 (current master), will open another RP

**Pull request closed without merging**